From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 20 Mar 2022 23:32:19 +1000
Subject: [PATCH] Add a timeout when saving chunks

This allows the server to exit cleanly instead of hanging forever
waiting for chunks to never finish loading.

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 7f5aa8f9cf6a32a7b2438d0d57b25f244e4db6b4..8d2b74a75d0ce130d76dd8c4ccdbbd3ee8440112 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -45,6 +45,7 @@ import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
 import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BooleanSupplier;
 import java.util.function.Consumer;
@@ -264,6 +265,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     private final ServerFunctionManager functionManager;
     private final FrameTimer frameTimer;
     private boolean enforceWhitelist;
+    public CompletableFuture<Void> saveTimeout = new CompletableFuture<>(); // MultiPaper
     private float averageTickTime;
     public final Executor executor;
     @Nullable
@@ -949,6 +951,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         }
 
         MinecraftServer.LOGGER.info("Saving worlds");
+        saveTimeout = saveTimeout.completeOnTimeout(null, 30, TimeUnit.SECONDS); // MultiPaper
         Iterator iterator = this.getAllLevels().iterator();
 
         ServerLevel worldserver;
diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index dc4d60f97b1a747116435b325946f1190398851f..3c59e28bf1d3139b2f93f707556519fa29bbc3d9 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -973,16 +973,20 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
                 mutableboolean.setFalse();
                 list.stream().map((playerchunk) -> {
                     CompletableFuture completablefuture;
+                    CompletableFuture timeoutable; // MultiPaper - add timeout
 
                     do {
                         completablefuture = playerchunk.getChunkToSave();
                         BlockableEventLoop iasynctaskhandler = this.mainThreadExecutor;
 
                         Objects.requireNonNull(completablefuture);
-                        iasynctaskhandler.managedBlock(completablefuture::isDone);
+                        timeoutable = CompletableFuture.anyOf(completablefuture, level.getServer().saveTimeout); // MultiPaper - add timeout
+                        iasynctaskhandler.managedBlock(timeoutable::isDone); // MultiPaper - add timeout
                     } while (completablefuture != playerchunk.getChunkToSave());
 
-                    return (ChunkAccess) completablefuture.join();
+                    if (!completablefuture.isDone()) LOGGER.warn("Timed out while waiting for chunk " + playerchunk.getWorld().getWorld().getName() + playerchunk.getPos() + " to finish loading, skipping saving it.");  // MultiPaper - add timeout
+
+                    return (ChunkAccess) timeoutable.join();  // MultiPaper - add timeout
                 }).filter((ichunkaccess) -> {
                     return ichunkaccess instanceof ImposterProtoChunk || ichunkaccess instanceof LevelChunk;
                 }).filter(this::save).forEach((ichunkaccess) -> {
